---
title: Flank length 100 - Compare distribution of bowtie alignment parameters between reads aligned correctly
  (TP) or not (FP)
output:
  html_document:
    df_print: paged
  html_notebook: default
  pdf_document: default
---


```{r}

library(ggplot2)
library(dplyr)
library(tidyr)

file=''
data<-read.delim(file, col.names=c("name", "qual", "AS", "XS", "XN", "XM", "XG", "NM", "MD", "BD", "Type"))

# Remove Filtered alignments
data <- data[data$BD!="Filtered",]

data$XS[is.nan(data$XS)] <- min(data$XS[! is.nan(data$XS)]) - 1
data$AS[is.nan(data$AS)] <- min(data$AS[! is.nan(data$AS)]) - 1

```

## Metrics distribution

### NM
NM:i:<N>
The edit distance; that is, the minimal number of one-nucleotide edits (substitutions, insertions and deletions) needed to transform the read string into the reference string. Only present if SAM record is for an aligned read.

```{r}

ggplot(data, aes(NM)) + geom_histogram() + facet_grid(Type~BD ) + scale_y_log10()

```

### AS
AS:i:<N>  
Alignment score. Can be negative. Can be greater than 0 in --local mode (but not in --end-to-end mode). Only present if SAM record is for an aligned read.


```{r}

ggplot(data, aes(AS)) + geom_histogram() + facet_grid(Type~BD ) + scale_y_log10()

```


### XS

XS:i:<N> 
Alignment score for the best-scoring alignment found other than the alignment reported. Can be negative. Can be greater than 0 in --local mode (but not in --end-to-end mode). Only present if the SAM record is for an aligned read and more than one alignment was found for the read. Note that, when the read is part of a concordantly-aligned pair, this score could be greater than AS:i.

```{r}

ggplot(data, aes(XS)) + geom_histogram() + facet_grid(Type~BD ) + scale_y_log10()

```



### qual

qual is the mapping quality

```{r}

ggplot(data, aes(qual)) + geom_histogram() + facet_grid(Type~BD ) + scale_y_log10()

```


## AS vs XS


```{r}

data %>% count(AS, XS, BD) -> count_AS_XS_BD

complete_count <- complete(count_AS_XS_BD, AS=seq(min(count_AS_XS_BD$AS), max(count_AS_XS_BD$AS)), XS=seq(min(count_AS_XS_BD$XS), max(count_AS_XS_BD$XS)), BD)

ggplot(complete_count, aes(x=AS, y=XS, fill=log(n))) + geom_tile() + facet_grid(~BD ) + scale_fill_gradient2(na.value="white")
```

## NM vs AS


```{r}

data %>% 
  count(AS, NM, BD) %>% 
  complete( AS=seq(min(data$AS), max(data$AS)), NM=seq(min(data$NM), max(data$NM)), BD) -> complete_count

ggplot(complete_count, aes(x=AS, y=NM, fill=log(n))) + geom_tile() + facet_grid(~BD ) + scale_fill_gradient2(na.value="white")

```



